### PR TITLE
openapi 3.0 rest descriptions

### PR DESCRIPTION
**Pull request by gmokki**

### CODE REVIEW

File: nflow-rest-api-common/src/main/java/io/nflow/rest/v1/msg/MaintenanceRequest.java [hunk 1]
```diff
@@ -4,35 +4,34 @@
...
   public MaintenanceRequestItem archiveWorkflows;
 
-  @ApiModelProperty("Delete passive workflows")
+  @Schema(description ="Delete passive workflows")
   public MaintenanceRequestItem deleteWorkflows;
 
   public static class MaintenanceRequestItem extends ModelObject {
-    @ApiModelProperty(value = "Workflow instances whose modified time is older than given period will be processed. Supports ISO-8601 format.", //
-        dataType = "String", example = "PT15D", required = true)
+    @Schema(description = "Workflow instances whose modified time is older than given period will be processed. Supports ISO-8601 format.", //
```
Comment (efonsell) [inline nflow-rest-api-common/src/main/java/io/nflow/rest/v1/msg/MaintenanceRequest.java:27]: Format is a sperate field now, so:
```suggestion
    @Schema(description = "Workflow instances whose modified time is older than given period will be processed",
```

File: nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/MaintenanceResource.java [hunk 2]
```diff
@@ -39,9 +46,10 @@
   private MaintenanceConverter converter;
 
   @POST
-  @ApiOperation(value = "Do maintenance on old workflow instances synchronously", response = MaintenanceResponse.class)
+  @Operation(description = "Do maintenance on old workflow instances synchronously")
+  @ApiResponse(responseCode = "200", description = "Maintenance operation status", content = @Content(schema = @Schema(implementation = MaintenanceResponse.class)))
   public Response cleanupWorkflows(
-      @ApiParam(value = "Parameters for the maintenance process", required = true) MaintenanceRequest request) {
+      @RequestBody(description = "Parameters for the maintenance process") MaintenanceRequest request) {
```
Comment (efonsell) [inline nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/MaintenanceResource.java:52]: is required=true the default?

Comment (efonsell) [reply nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/MaintenanceResource.java:52]: Should there be @Valid annotation here as well?

File: nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java [hunk 3]
```diff
@@ -164,9 +176,10 @@ public Response setSignal(@ApiParam("Internal id for workflow instance") @PathPa
 
   @PUT
   @Path("/{id}/wakeup")
-  @ApiOperation(value = "Wake up sleeping workflow instance", response = WakeupResponse.class, notes = "If expected states are given, only wake up if the instance is in one of the expected states.")
-  public Response wakeup(@ApiParam("Internal id for workflow instance") @PathParam("id") long id,
-      @Valid @ApiParam("Expected states") WakeupRequest req) {
+        @Operation(description = "Wake up sleeping workflow instance. If expected states are given, only wake up if the instance is in one of the expected states.")
```
Comment (efonsell) [inline nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java:179]: Put notes to description like above instead of merging to description?

File: nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java [hunk 4]
```diff
@@ -79,11 +87,11 @@ public WorkflowInstanceResource(SchedulerService scheduler, WorkflowInstanceServ
...
   @PutMapping(consumes = APPLICATION_JSON_VALUE)
-  @ApiOperation(value = "Submit new workflow instance")
-  @ApiResponses({ @ApiResponse(code = 201, message = "Workflow was created", response = CreateWorkflowInstanceResponse.class),
-      @ApiResponse(code = 400, message = "If instance could not be created, for example when state variable value was too long") })
+  @Operation(summary = "Submit new workflow instance")
+  @ApiResponses({ @ApiResponse(responseCode = "201", description = "Workflow was created", content = @Content(schema = @Schema(implementation = CreateWorkflowInstanceResponse.class))),
+          @ApiResponse(responseCode = "400", description = "If instance could not be created, for example when state variable value was too long") })
   public Mono<ResponseEntity<?>> createWorkflowInstance(
-      @RequestBody @ApiParam(value = "Submitted workflow instance information", required = true) CreateWorkflowInstanceRequest req) {
+          @RequestBody @Parameter(description = "Submitted workflow instance information", required = true) CreateWorkflowInstanceRequest req) {
```
Comment (efonsell) [inline nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java:94]: @Valid

File: nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java [hunk 5]
```diff
@@ -93,60 +101,62 @@ public WorkflowInstanceResource(SchedulerService scheduler, WorkflowInstanceServ
...
-      @ApiResponse(code = 400, message = "If instance could not be updated, for example when state variable value was too long"),
-      @ApiResponse(code = 409, message = "If workflow was executing and no update was done") })
-  public Mono<ResponseEntity<?>> updateWorkflowInstance(
-      @ApiParam("Internal id for workflow instance") @PathVariable("id") long id,
-      @RequestBody @ApiParam("Submitted workflow instance information") UpdateWorkflowInstanceRequest req) {
+  @ApiResponses({ @ApiResponse(responseCode = "204", description = "If update was successful"),
+          @ApiResponse(responseCode = "400", description = "If instance could not be updated, for example when state variable value was too long"),
+          @ApiResponse(responseCode = "409", description = "If workflow was executing and no update was done") })
+  public Mono<ResponseEntity<?>> updateWorkflowInstance(@Parameter(description = "Internal id for workflow instance") @PathVariable("id") long id,
+                                                  @RequestBody @Parameter(description = "Submitted workflow instance information") UpdateWorkflowInstanceRequest req) {
```
Comment (efonsell) [inline nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java:110]: @Valid

File: nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java [hunk 6]
```diff
@@ -93,60 +101,62 @@ public WorkflowInstanceResource(SchedulerService scheduler, WorkflowInstanceServ
...
   @SuppressFBWarnings(value = "LEST_LOST_EXCEPTION_STACK_TRACE", justification = "The empty result exception contains no useful information")
-  public Mono<ResponseEntity<?>> fetchWorkflowInstance(@ApiParam("Internal id for workflow instance") @PathVariable("id") long id,
-      @RequestParam(value = "include", required = false) @ApiParam(value = INCLUDE_PARAM_DESC, allowableValues = INCLUDE_PARAM_VALUES, allowMultiple = true) String include,
-      @RequestParam(value = "queryArchive", required = false, defaultValue = QUERY_ARCHIVED_DEFAULT_STR) @ApiParam("Query also the archive if not found from main tables") boolean queryArchive,
-      @RequestParam(value = "maxActions", required = false) @ApiParam("Maximum number of actions returned for each workflow instance") Long maxActions) {
+  public Mono<ResponseEntity<?>> fetchWorkflowInstance(
+          @Parameter(description = "Internal id for workflow instance") @PathVariable("id") long id,
+          @RequestParam(value = "include", required = false) @Parameter(description = INCLUDE_PARAM_DESC/*, allowableValues = INCLUDE_PARAM_VALUES, allowMultiple = true*/) String include,
```
Comment (efonsell) [inline nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java:124]: what to do with allowableValues?

File: nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java [hunk 7]
```diff
@@ -116,44 +125,47 @@ public Response updateWorkflowInstance(@ApiParam("Internal id for workflow insta
...
-  public Response fetchWorkflowInstance(@ApiParam("Internal id for workflow instance") @PathParam("id") long id,
-      @QueryParam("include") @ApiParam(value = INCLUDE_PARAM_DESC, allowableValues = INCLUDE_PARAM_VALUES, allowMultiple = true) String include,
-      @QueryParam("maxActions") @ApiParam("Maximum number of actions returned for each workflow instance") Long maxActions,
-      @QueryParam("queryArchive") @ApiParam("Query also the archive if not found from main tables") Boolean queryArchive) {
+  @ApiResponses({ @ApiResponse(responseCode = "200", description = "", content = @Content(schema = @Schema(implementation = ListWorkflowInstanceResponse.class))),
+          @ApiResponse(responseCode = "404", description = "If instance could not be created, for example when state variable value was too long") })
+  public Response fetchWorkflowInstance(@Parameter(description = "Internal id for workflow instance") @PathParam("id") long id,
+          @QueryParam("include") @Parameter(description = INCLUDE_PARAM_DESC/*, allowableValues = INCLUDE_PARAM_VALUES, allowMultiple = true*/) String include,
```
Comment (efonsell) [inline nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java:133]: what to do with allowableValues?

**Review by efonsell — COMMENTED** [review]
Minor comments

File: nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ResourceBase.java [hunk 8]
```diff
@@ -149,24 +149,15 @@ public boolean updateWorkflowInstance(final long id,
       String stateVariableKey, String stateVariableValue, String include, Long maxResults, Long maxActions, boolean queryArchive,
       WorkflowInstanceService workflowInstances, ListWorkflowInstanceConverter listWorkflowConverter) {
     Set<String> includeStrings = parseIncludeStrings(include).collect(toSet());
-    QueryWorkflowInstances q = new QueryWorkflowInstances.Builder()
-        .addIds(ids.toArray(new Long[ids.size()]))
-        .addTypes(types.toArray(new String[types.size()]))
-        .setParentWorkflowId(parentWorkflowId)
-        .setParentActionId(parentActionId)
+    QueryWorkflowInstances q = new QueryWorkflowInstances.Builder().addIds(ids.toArray(new Long[ids.size()]))
+        .addTypes(types.toArray(new String[types.size()])).setParentWorkflowId(parentWorkflowId).setParentActionId(parentActionId)
```
Comment (gmokki) [inline nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ResourceBase.java:153]: Should we add `//` after each method to keep them on separate lines?

Comment (efonsell) [reply nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ResourceBase.java:153]: nope, I just need to fix the formatter settings

File: nflow-rest-api-jax-rs/src/test/java/io/nflow/rest/v1/jaxrs/WorkflowDefinitionResourceTest.java [hunk 9]
```diff
@@ -116,7 +115,7 @@ public void listWorkflowDefinitionsFindsExistingAndStoredDefinitionsWithDbType()
 
   private List<ListWorkflowDefinitionResponse> getDefinitionListType(List<String> list) {
     try (Response r = resource.listWorkflowDefinitions(list)) {
-      return r.readEntity(definitionListType);
+      return asList(r.readEntity(ListWorkflowDefinitionResponse[].class));
```
Comment (gmokki) [inline nflow-rest-api-jax-rs/src/test/java/io/nflow/rest/v1/jaxrs/WorkflowDefinitionResourceTest.java:118]: Why this change? First read as array. Then copy it into a list rather than directly add items to a list - not that there will be so many definitions that it the performance difference would matter.

Comment (efonsell) [reply nflow-rest-api-jax-rs/src/test/java/io/nflow/rest/v1/jaxrs/WorkflowDefinitionResourceTest.java:118]: Just to get rid of the ugly GenericType definitions. Anyway this is just test code. Maybe this method could return an array?

Comment (efonsell) [reply nflow-rest-api-jax-rs/src/test/java/io/nflow/rest/v1/jaxrs/WorkflowDefinitionResourceTest.java]: Reverted

File: nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java [hunk 10]
```diff
@@ -116,44 +129,61 @@ public Response updateWorkflowInstance(@ApiParam("Internal id for workflow insta
...
+      description = "Fetch full state and action history of a single workflow instance.")
+  @SuppressFBWarnings(value = "LEST_LOST_EXCEPTION_STACK_TRACE",
+      justification = "The empty result exception contains no useful information")
+  @ApiResponses({
+      @ApiResponse(responseCode = "200", description = "",
+          content = @Content(schema = @Schema(implementation = ListWorkflowInstanceResponse.class))),
+      @ApiResponse(responseCode = "404",
+          description = "If instance could not be created, for example when state variable value was too long") })
+  public Response fetchWorkflowInstance(@Parameter(description = "Internal id for workflow instance") @PathParam("id") long id,
+      @QueryParam("include") @Parameter(description = INCLUDE_PARAM_DESC) String include,
```
Comment (efonsell) [inline nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java:142]: Pitäiskö tämä muuttaa tästä "include=include1,include,include3" formaatista muotoon "include=include1&include=include2&include=include3" tms?

Comment (gmokki) [reply nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java:142]: Personally I think each include in separate parameter is more classic style. But some modern frameworks do not support it easily anymore.
So I would support both, but only document the multiple repeated parameters.
If we get `String[] include` as parameter we can do
```Stream.of(include).flatMap(i -> Stream.of(i.split(",")).collect(toSet())``` to flatten both cases into a set that can be later queried fast.

Comment (efonsell) [reply nflow-rest-api-jax-rs/src/main/java/io/nflow/rest/v1/jaxrs/WorkflowInstanceResource.java]: Old format is deprecated but still supported by `include` param.
Added support for new format with `includes` param.

**Review by efonsell — APPROVED** [review]
IMHO this is now ok to merge. More improvements can be done before 8.0.0 release if necessary.

File: nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java [hunk 11]
```diff
@@ -93,60 +103,94 @@ public WorkflowInstanceResource(SchedulerService scheduler, WorkflowInstanceServ
...
-      @RequestParam(value = "include", required = false) @ApiParam(value = INCLUDE_PARAM_DESC, allowableValues = INCLUDE_PARAM_VALUES, allowMultiple = true) String include,
-      @RequestParam(value = "queryArchive", required = false, defaultValue = QUERY_ARCHIVED_DEFAULT_STR) @ApiParam("Query also the archive if not enough results found from main tables") boolean queryArchive,
-      @RequestParam(value = "maxResults", required = false) @ApiParam("Maximum number of workflow instances to be returned") Long maxResults,
-      @RequestParam(value = "maxActions", required = false) @ApiParam("Maximum number of actions returned for each workflow instance") Long maxActions) {
-    return handleExceptions(() -> wrapBlocking(
-        () -> ok(super.listWorkflowInstances(ids, types, parentWorkflowId, parentActionId, states, statuses, businessKey,
-            externalId, stateVariableKey, stateVariableValue, include, maxResults, maxActions, queryArchive, this.workflowInstances,
-            this.listWorkflowConverter).iterator())));
+      @RequestParam(value = "id", defaultValue = "") @Parameter(description = "Internal id of workflow instance") Set<Long> ids,
```
Comment (gmokki) [inline nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java:152]: Most of the other RequestParams have required=false, but id does not have? does the defaultValue imply required=false?

File: nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ResourceBase.java [hunk 12]
```diff
@@ -158,35 +157,42 @@ public boolean updateWorkflowInstance(final long id,
...
-  private Set<WorkflowInstanceInclude> parseIncludeEnums(String include) {
-    return parseIncludeStrings(include).map(INCLUDE_STRING_TO_ENUM::get).filter(Objects::nonNull)
-        .collect(toCollection(HashSet::new));
+    return instances.map(instance -> listWorkflowConverter.convert(instance, propertyIncludes, queryArchive));
   }
 
-  private Stream<String> parseIncludeStrings(String include) {
-    return Stream.of(trimToEmpty(include).split(","));
+  private Set<ApiWorkflowInstanceInclude> resolveIncludes(Set<ApiWorkflowInstanceInclude> includes, String include) {
+    Set<ApiWorkflowInstanceInclude> allIncludes = new HashSet<>();
```
Comment (gmokki) [inline nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ResourceBase.java:174]: `EnumSet.noneOf(ApiWorkflowInstanceInclude.class)`

File: nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ApiWorkflowInstanceInclude.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    return include;
+  }
+
+  /**
+   * Resolve enum value from string.
+   *
+   * @return Matching enum value, or null if no match found.
+   */
+  public static ApiWorkflowInstanceInclude fromValue(String value) {
+    return stream(values()).filter(include -> include.name().equals(value)).findAny().orElse(null);
```
Comment (gmokki) [inline nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ApiWorkflowInstanceInclude.java:53]: I think the normal and fast way of writing this is:
```java
try {
  return valueOf(value);
} catch (IllegalArgumentException ex) {
  return null;
}
```

File: nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ResourceBase.java [hunk 14]
```diff
@@ -158,35 +157,42 @@ public boolean updateWorkflowInstance(final long id,
...
+    if (includes != null) {
+      allIncludes.addAll(includes);
+    }
+    if (isNotBlank(include)) {
+      allIncludes.addAll(stream(include.split(include))
```
Comment (gmokki) [inline nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ResourceBase.java:179]: No reason to first collect a set and then add to another. PS the split was buggy. I guess you did not leave any tests for it?

How about:
```java
stream(include.split(","))
          .map(ApiWorkflowInstanceInclude::fromValue)
          .filter(Objects::nonNull)
          .forEach(allIncludes::add);
```

File: nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java (hunk 11 above)
Comment (efonsell) [reply nflow-rest-api-spring-web/src/main/java/io/nflow/rest/v1/springweb/WorkflowInstanceResource.java]: fixed

File: nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ApiWorkflowInstanceInclude.java (hunk 13 above)
Comment (efonsell) [reply nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ApiWorkflowInstanceInclude.java]: changed

File: nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ResourceBase.java (hunk 12 above)
Comment (efonsell) [reply nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ResourceBase.java]: changed

File: nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ResourceBase.java (hunk 14 above)
Comment (efonsell) [reply nflow-rest-api-common/src/main/java/io/nflow/rest/v1/ResourceBase.java]: fixed
